From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Da Xue <da@libre.computer>
Date: Thu, 21 Jul 2022 18:08:21 -0400
Subject: [PATCH] mmc: increase stabilization delay from 1ms to 20ms

Some users experienced problems booting u-boot from SPL hanging here:

Trying to boot from MMC1 or Trying to boot from MMC2

This seems to occur with both MicroSD and eMMC modules on ALL-H3-CC.
Increasing the delay after mmc reset fixes these boot problems.
Some MicroSD cards are impacted more than others so it is possible that
MicroSD internals need time to stabilize. Below is some failure data.

sandisk ultra   64gb 9/20 with 1ms,  20/20 with 10ms
sandisk ultra   16gb 2/20 with 1ms,  20/20 with 10ms
sandisk extreme 16gb 6/20 with 10ms, 20/20 with 20ms

A quick comparison of schematics show series resistors for ESD
protection on the MicroSD GPIOs not present on all H3/H5 boards.
It is not known if this is related to the issue.

This patch adds a fixed 20ms delay to mmc init to mitigate the problem.
If boot time optimization is required and the platform does not require
the delay. The delay can be replaced with:

writel(SUNXI_MMC_GCTRL_RESET, &priv->reg->gctrl);
if (wait_for_bit_le32( &priv->reg->gctrl,
SUNXI_MMC_GCTRL_RESET, false, 20, false)) {
printf("%s: Timeout\n", __func__);
return -ETIMEDOUT;
}

Signed-off-by: Da Xue <da@libre.computer>
---
 drivers/mmc/sunxi_mmc.c | 24 ++++++++++++++++--------
 1 file changed, 16 insertions(+), 8 deletions(-)

--- a/drivers/mmc/sunxi_mmc.c
+++ b/drivers/mmc/sunxi_mmc.c
@@ -285,9 +285,10 @@
 
 	for (i = 0; i < word_cnt;) {
 		unsigned int in_fifo;
+		unsigned int loop_count = 10000000;
 
 		while ((status = readl(&priv->reg->status)) & status_bit) {
-			if (get_timer(start) > timeout_msecs)
+			if (get_timer(start) > timeout_msecs || --loop_count == 0)
 				return -1;
 		}
 
@@ -316,9 +317,11 @@
 		in_fifo = SUNXI_MMC_STATUS_FIFO_LEVEL(status);
 		if (in_fifo == 0 && (status & SUNXI_MMC_STATUS_FIFO_FULL))
 			in_fifo = 32;
+		if (in_fifo > word_cnt - i)
+			in_fifo = word_cnt - i;
 		for (; in_fifo > 0; in_fifo--)
 			buff[i++] = readl_relaxed(&priv->reg->fifo);
-		dmb();
+		mb();
 	}
 
 	return 0;
@@ -470,20 +474,20 @@
 {
 	/* Reset controller */
 	writel(SUNXI_MMC_GCTRL_RESET, regs + SUNXI_MMC_GCTRL);
+	udelay(20000);
+
+	/* Reset card on all SoCs supporting eMMC hardware reset */
+	writel(SUNXI_MMC_HWRST_ASSERT, regs + SUNXI_MMC_HWRST);
+	udelay(50);
+	writel(SUNXI_MMC_HWRST_DEASSERT, regs + SUNXI_MMC_HWRST);
 	udelay(1000);
 
 	if (IS_ENABLED(CONFIG_SUN50I_GEN_H6) || IS_ENABLED(CONFIG_SUNXI_GEN_NCAT2)) {
-		/* Reset card */
-		writel(SUNXI_MMC_HWRST_ASSERT, regs + SUNXI_MMC_HWRST);
-		udelay(10);
-		writel(SUNXI_MMC_HWRST_DEASSERT, regs + SUNXI_MMC_HWRST);
-		udelay(300);
-
 		/* Setup FIFO R/W threshold. Needed on H616. */
 		writel(SUNXI_MMC_THLDC_READ_THLD(512) |
 		       SUNXI_MMC_THLDC_WRITE_EN |
 		       SUNXI_MMC_THLDC_READ_EN, regs + SUNXI_MMC_THLDC);
 	}
 }
 
 /* non-DM code here is used by the (ARM) SPL only */
